Added the Context,Container and Component diagrams

MadhuRoy-git opened the pull request.

MadhuRoy-git committed:
Added the Context,Container and Component diagrams

hugh-emerson reviewed (approved):
Overall you've got the right idea. But here are some possible improvements:

For the container diagram, it looks like you took an example from somewhere, but it isn't quite true for your web app. "Single page application" refers to web apps where there the browser only loads one page and then uses javascript to update it as the user interacts with it, making HTTP requests to the server as needed, but not loading a new html page from the server.

In contrast your app is sending off a full HTML page to the user each time they make a request. 

Even if it were a single page app, I don't think it's true to draw your diagram as if two separate things are served via gunicorn. The user makes requests to gunicorn, which forwards it to the Flask app and also forwards responses from the flask app to the user.

Also you should make it clear what you are zooming in on, by putting those bits inside a large dotted box, for example. So that it's still clear that e.g. Trello is some external service where you don't care about any internal details
```
user --[->gunicorn -> flask app-]--> Trello
```

The component diagram should also be changed slightly (see attached image for my quick version of these suggestions):
- It shouldn't say "single page app", it should say it returns templated responses (using your template html files and the Jinja templating engine if you want to include the details).
- You need to make the entrypoint obvious and also make it obvious that Trello is some external service
- Also, note that you said that requests are made by using buttons etc on the page but just visiting the page in the first place by typing the URL in the browser will also make a request to Flask.

![image](https://user-images.githubusercontent.com/10462681/102885973-25f5c880-444c-11eb-80c4-d5b4110ded00.png)

MadhuRoy-git committed:
Review comments addressed

MadhuRoy-git commented:
> Overall you've got the right idea. But here are some possible improvements:
> 
> For the container diagram, it looks like you took an example from somewhere, but it isn't quite true for your web app. "Single page application" refers to web apps where there the browser only loads one page and then uses javascript to update it as the user interacts with it, making HTTP requests to the server as needed, but not loading a new html page from the server.
> 
> In contrast your app is sending off a full HTML page to the user each time they make a request.
> 
> Even if it were a single page app, I don't think it's true to draw your diagram as if two separate things are served via gunicorn. The user makes requests to gunicorn, which forwards it to the Flask app and also forwards responses from the flask app to the user.
> 
> Also you should make it clear what you are zooming in on, by putting those bits inside a large dotted box, for example. So that it's still clear that e.g. Trello is some external service where you don't care about any internal details
> 
> ```
> user --[->gunicorn -> flask app-]--> Trello
> ```
> 
> The component diagram should also be changed slightly (see attached image for my quick version of these suggestions):
> 
> * It shouldn't say "single page app", it should say it returns templated responses (using your template html files and the Jinja templating engine if you want to include the details).
> * You need to make the entrypoint obvious and also make it obvious that Trello is some external service
> * Also, note that you said that requests are made by using buttons etc on the page but just visiting the page in the first place by typing the URL in the browser will also make a request to Flask.
> 
> ![image](https://user-images.githubusercontent.com/10462681/102885973-25f5c880-444c-11eb-80c4-d5b4110ded00.png)

MadhuRoy-git commented:
@hugh-emerson , I have addressed your PR comments and pushed the changes to the branch. Let me know if you're happy for me to merge after these changes.
